distributed/scheduler.py

```diff
@@ -69,6 +70,242 @@
+            assert isinstance(cs, ClientState), (repr(cs), self.who_wants)
+        for ws in self.who_has:
+            assert isinstance(ws, WorkerState), (repr(ws), self.who_has)
+        for ts in self.dependencies:
+            assert isinstance(ts, TaskState), (repr(ts), self.dependencies)
+        for ts in self.dependents:
+            assert isinstance(ts, TaskState), (repr(ts), self.dependents)
+
+
+class _StateLegacyMapping(Mapping):
+    """
+    A mapping interface mimicking the former Scheduler state dictionaries.
+    """
+
+    def __init__(self, states, accessor):
+        self._states = states
+        self._accessor = accessor
+
+    def __iter__(self):
+        return iter(self._states)
+
+    def __len__(self):
+        return len(self._states)
+
+    def __getitem__(self, key):
+        return self._accessor(self._states[key])
+
+    def __repr__(self):
+        return "%s(%s)" % (self.__class__, dict(self))
+
+
+class _OptionalStateLegacyMapping(_StateLegacyMapping):
+    """
+    Similar to _StateLegacyMapping, but a false-y value is interpreted
+    as a missing key.
+    """
+    # For tasks etc.
+
+    def __iter__(self):
+        accessor = self._accessor
+        for k, v in self._states.items():
+            if accessor(v):
+                yield k
+
+    def __len__(self):
+        accessor = self._accessor
+        return sum(bool(accessor(v)) for v in self._states.values())
+
+    def __getitem__(self, key):
+        v = self._accessor(self._states[key])
+        if v:
+            return v
+        else:
+            raise KeyError
+
+
+class _StateLegacySet(Set):
+    """
+    Similar to _StateLegacyMapping, but exposes a set containing
+    all values with a true value.
+    """
+    # For loose_restrictions
+
+    def __init__(self, states, accessor):
+        self._states = states
+        self._accessor = accessor
+
+    def __iter__(self):
+        return (k for k, v in self._states.items() if self._accessor(v))
+
+    def __len__(self):
+        return sum(map(bool, map(self._accessor, self._states.values())))
+
+    def __contains__(self, k):
+        st = self._states.get(k)
+        return st is not None and bool(self._accessor(st))
+
+    def __repr__(self):
+        return "%s(%s)" % (self.__class__, set(self))
+
+
+def _legacy_task_key_set(tasks):
+    """
+    Transform a set of task states into a set of task keys.
+    """
+    return {ts.key for ts in tasks}
+
+
+def _legacy_client_key_set(clients):
+    """
+    Transform a set of client states into a set of client keys.
+    """
+    return {cs.client_key for cs in clients}
+
+
+def _legacy_worker_key_set(workers):
+    """
+    Transform a set of worker states into a set of worker keys.
+    """
+    return {ws.worker_key for ws in workers}
+
+
+def _legacy_task_key_dict(task_dict):
+    """
+    Transform a dict of {task state: value} into a dict of {task key: value}.
+    """
+    return {ts.key: value for ts, value in task_dict.items()}
```

Comment:
I suppose that these will eventually go away as you work through more of the peripheral modules?

Reply:
Ideally yes. Some though are used a lot in tests, so that will require a fair bit of manual fixing (unless there's a way to automate the changes).